### ISSUE TITLE
[backend] Create an incremental backup script that doesn't freeze SQL

### ISSUE THREAD

**Issue by fabd** [fabd/kanji-koohii#69]
Disclaimer: I love UX and frontend, so my backend / SQL skills are lacking.

Currently I am using a CRON job to make daily backups. I had to set compression level close to the lowest (which is still very good for SQL text).. otherwise the server was killing the thread due to taking too much time.

As of writing, it takes a little over a minute to create a `.tgz` backup of the entire Kanji Koohii database. It is convenient to have a single file. However, the mysqldump + archive **freezes the server for about a minute**.

This is bad for users and just bad for my backups too.

## Solution 1 : simple? But how?

I have to find some kind of incremental backup. I don't know if I can do that with a long command line alone with `mysqldump`? Potentially query all the tables based on last edit timestamp to output only the rows from the last week, for example. This would drastically reduce the time it takes to do a backup as well as the archive size obviously, although it would require careful consideration of the indexes on the tables (so the rows can be selected by timestamp fast).

This could be a simple and satisfying solution **IF** and only if I can just overwrite the last full backup with the incremental ones. Eg. do MySQL `source  incremental-backup.sql` one after the other. Otherwise I would have something complicated to restore in case of emergency as well as duplication for testing locally.

This also assumes I still do a full backup but only once a week for example. 

## Solution 2 : use a custom backup script

So I know forums like MyBB have a "task" system. Well is there a generic solution out there for MySQL backups? It sounds like a problem someone already solved and that doesn't need to be too specific to an app. You'd tell the script what your schema looks like, and it would do the above, based on criteria like how far back to save (eg. last week), the naming of the output SQL files, etc.

Ideally such a script wouldbe able to piece back together a full snapshot of the database from the incremental backups so you don't have to manually piece it back together.

## Specs

I'm afraid the backend uses Symfony 1.x but doesn't use the built-in ORMs for legacy reasons. The website has seen two big refactors in the past 11 years, and it's currently using something very close to the Zend Db API so a solution based on that is best. Although for a backup script just safely built `mysql_xyz` calls is fine too. 

It's not super urgent. Any advice from people with SQL experience very much appreciated. As well as recommendations for php packages I could look into?

**Comment by HelenFoster** [comment]
It may be possible to get big improvements by tweaking the mysqldump command.

The `--quick` option makes the dump more efficient, but is quite likely enabled already.

With InnoDB, the `--single-transaction` option allows the website to keep working while the backup is in progress, rather than locking the tables.

If the server is still overloaded with these options, it may be possible to slow down the backup, so it takes longer but uses a lower proportion of the resources at any one moment (only for InnoDB single-transaction). `nice` doesn't help with the actual dumping, because the work for that is done by the server process, but it might be useful if too much CPU time is going to compression. This person claims to have reduced the dumping resource use by limiting the bandwidth with `cstream` : http://serverfault.com/a/313232

**Comment by fabd** [comment]
Thank you @HelenFoster 

I think the CPU usage is not a concern. But I pipe the mysqldump into gzip and I think that's the part which is slow. The thread gets killed if it takes too long.

Here is what a CRON job looks like:

```bash
mysqldump --opt -u user -ppassword db_name | gzip -v1 - > `/bin/date +\%F`-db_name.sql.gz
```

I just realized I could potentially split those in successive CRON jobs so the gzip is done in a second step but I'd rather keep it as simple as I can.

**Comment by HelenFoster** [comment]
Indeed, the backup is locking the database, so the website can't do anything until the backup is finished.

Your idea of doing the gzip in a second step would reduce the time for which the database is locked.

Does the large InnoDB story table need to be perfectly consistent with the others? I guess the story table has references to the MyISAM tables; if there are no references the other way, and the sort of things the story table refers to don't get deleted, then backing up the story table first might be an option.

**Comment by fabd** [comment]
Note to self:

- split mysql export to backup the two largest tables (stories and flashcards) on separate days, alternating
